base/reducedim.jl

```diff
@@ -250,44 +135,175 @@ _firstslice(i::OneTo) = OneTo(1)
 _firstslice(i::Slice) = Slice(_firstslice(i.indices))
 _firstslice(i) = i[firstindex(i):firstindex(i)]
 
-function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted)
+function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted, Aax=axes(A))
     lsiz = check_reducedims(R,A)
     isempty(A) && return R
 
-    if has_fast_linear_indexing(A) && lsiz > 16
+    if has_fast_linear_indexing(A) && lsiz > 16 && axes(A) === Aax
         # use mapreduce_impl, which is probably better tuned to achieve higher performance
-        nslices = div(length(A), lsiz)
         ibase = first(LinearIndices(A))-1
-        for i = 1:nslices
+        for i in eachindex(R) # TODO: add tests for this change
             @inbounds R[i] = op(R[i], mapreduce_impl(f, op, A, ibase+1, ibase+lsiz))
             ibase += lsiz
         end
         return R
     end
-    indsAt, indsRt = safe_tail(axes(A)), safe_tail(axes(R)) # handle d=1 manually
+    indsAt, indsRt = safe_tail(Aax), safe_tail(axes(R)) # handle d=1 manually
     keep, Idefault = Broadcast.shapeindexer(indsRt)
     if reducedim1(R, A)
         # keep the accumulator as a local variable when reducing along the first dimension
         i1 = first(axes1(R))
         @inbounds for IA in CartesianIndices(indsAt)
             IR = Broadcast.newindex(IA, keep, Idefault)
             r = R[i1,IR]
-            @simd for i in axes(A, 1)
-                r = op(r, f(A[i, IA]))
+            @simd for i in Aax[1]
+                r = op(r, f(A[i, IA])) # this could also be done pairwise
             end
             R[i1,IR] = r
         end
     else
         @inbounds for IA in CartesianIndices(indsAt)
             IR = Broadcast.newindex(IA, keep, Idefault)
-            @simd for i in axes(A, 1)
+            @simd for i in Aax[1]
                 R[i,IR] = op(R[i,IR], f(A[i,IA]))
             end
         end
     end
     return R
 end
 
+if isdefined(Core, :Compiler)
+    function _mapreduce_might_widen(f::F, op::G, ::Type{T}) where {F,G,T}
+        return !isconcretetype(Core.Compiler.return_type(x->op(f(x), f(x)), Tuple{T}))
+    end
+else
+    _mapreduce_might_widen(_, _, _) = true
+end
+
+
+
+#### Reduction initialization ####
+#
+# Similar to the scalar mapreduce, there are two*three cases we need to consider
+# when constructing the initial results array:
+# * Have init: Fill results array with init, proceed (regardless of reduction size)
+#   * Don't worry about widening and do a straight iteration with in-place updates over all values
+# * Otherwise:
+#   * No values: Fill with `mapreduce_empty` & return
+#   * One value: Fill with `mapreduce_first` & return
+#   * Two+ values: Fill with `op(f(a1), f(a2))`, proceed <--- this is the hard case:
+#       * Proceeding involves skipping the first two inner iterations
+_mapreduce_dim(::Type{F}, op::G, ::_InitialValue, A::AbstractArrayOrBroadcasted, dims) where {F, G} = _mapreduce_dim(x->F(x), op, Base._InitialValue(), A, dims)
```

Comment:
I needed this to get `maximum(Float64, reshape(1:4, 2, :); dims = 2)` to infer.  Something something specialization heuristics.